### PR TITLE
Add dataset/sampler benchmarking script

### LINKED ISSUE #81: Benchmarking of GeoDataset for a paper result
### Datasets

We want to test several popular image sources, as well as both raster and vector labels.

- [ ] NAIP + Chesapeake
- [ ] Landsat + CDL
- [ ] Sentinel + Canadian Building Footprints

There is also a question of which file formats to test. For example, sampling from GeoJSON can take 3 min per getitem, whereas ESRI Shapefile only takes 1 sec per getitem (https://github.com/microsoft/torchgeo/pull/69#issuecomment-892816306).

### Experiments

- [ ] **Data location:** local vs. blob storage
- [ ] **Sampling strategy:** random vs. 2-step random tile/chip vs. non-random grid sampling
- [ ] **Warping strategy:** on-the-fly vs. pre-processing step
- [ ] **I/O strategy:** load/warp entire file vs. load/warp/cache entire file vs. load/warp single window

For the warping strategy, we should test the following possibilities:

- [ ] Already in correct CRS/resolution
- [ ] Need to change CRS
- [ ] Need to change resolution
- [ ] Need to change both CRS and resolution

What is the upfront cost of these pre-processing steps?

Example notebook: https://gist.github.com/calebrob6/d9bc5609ff638d601e2c35a1ab0a2dec

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR adds a script to benchmark our GeoDataset/GeoSampler implementations. I would like to commit this script to the repo for three reasons:

1. Reproducibility: if someone wants to reproduce the results in our paper, it should be easy
2. Regression: we should run these benchmarks again before each release to make sure sampling hasn't gotten slower
3. Refinement: if someone submits a PR to improve performance, we should have a well-defined metric for comparison

Now I just need to download some data and run it. Before I run it, we should make sure these default sizes/CRS/res/etc. look reasonable.

Closes #81

### CODE REVIEW

**Comment by calebrob6** [comment]
Some fun stuff I found while trying to get this to work:
- The CDL dataset will think `/datadrive/cdl` is corrupt -- https://github.com/microsoft/torchgeo/issues/99
- This line is mega broken https://github.com/microsoft/torchgeo/blob/paper/benchmarks/torchgeo/datasets/geo.py#L234
  - Specifically, this adds each tiff to the index occupying a single point in time. Now, the index will only return hits if a query has _exactly_ the same timestamp for mint and maxt. The _bounds_ of the index will usually look valid, but you'll never get results back (e.g. this is what happens in benchmark.py).
  - A simple workaround if you just want to benchmark stuff is to make maxt = mint + 1 year
- I don't think the RandomGeoSampler is going to work

The ZipDataset looks to be working after the hack though:

image

**Comment by calebrob6** [comment]
Committed a few things and the script works now

benchmark.py notes:
- I added a way to specify the total number of samples you want to load OR the total number of batches you want to load mainly for convenience.
- Note: previously, `len(samples)` was used in the dataloader to count the number of patches, but this returns the number of keys in the dict, not the batch size.
- We need to add a way to save results to file (appending to a CSV with column headers for all the args is sufficient, no need to over-engineer)
- I was wrong about the RandomGeoSampler -- it works totally fine!

**Comment by adamjstewart** [comment]
Undid some of the hacks and opened up separate PRs to fix them (#134, #138). Once those are merged I'll rebase this PR and start doing some benchmarking.

**Comment by adamjstewart** [comment]
Most recent commit fixes a couple of bugs I found:

* We were running the GridGeoSampler for N+1 batches instead of N batches
* We were using the same number of total patches even if the GridGeoSampler finished early

**Comment by adamjstewart** [comment]
Found another bug, we're reusing the same dataset for all samplers. For the first sampler, the cache is empty, but for later samplers it is not. So this results in an unfair comparison between samplers.

Which of the following options should we use?

1. Create a new dataset for each sampler
2. Run some random sampler first before testing all 3 samplers

Option 1 represents the average sampling rate when starting from scratch with an empty cache. Option 2 represents a better average after the cache has already been populated. I'm leaning towards 2 since the cache will be full for the vast majority of the training time when running through multiple epochs.

**Comment by adamjstewart** [comment]
We can also manually clear the cache between samplers: https://www.geeksforgeeks.org/clear-lru-cache-in-python/

File: benchmark.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        "cuda" if torch.cuda.is_available() else "cpu"
+    )
+    model = model.to(device)
+
+    tic = time.time()
+    num_total_patches = 0
+    for _ in range(num_batches):
+        num_total_patches += args.batch_size
+        x = torch.rand(args.batch_size, len(bands), args.patch_size, args.patch_size)
+        # y = torch.randint(0, 256, (args.batch_size, args.patch_size, args.patch_size))
```
Comment (adamjstewart) [inline benchmark.py:244]: This line is left over from when I was thinking about benchmarking a segmentation model instead of ResNet. I think segmentation is one of the more common tasks in remote sensing, and the models are more complex and therefore slower. If we want to have a slower model for comparison with our data loading rates, it might be good to use something like Mask R-CNN instead.

File: benchmark.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    model.conv1 = nn.Conv2d(  # type: ignore[attr-defined]
+        len(bands), 64, kernel_size=7, stride=2, padding=3, bias=False
+    )
+
+    criterion = nn.CrossEntropyLoss()  # type: ignore[attr-defined]
+    params = model.parameters()
+    optimizer = optim.SGD(params, lr=0.0001)
+
+    device = torch.device(  # type: ignore[attr-defined]
+        "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (calebrob6) [inline benchmark.py:235]: Would be good to specify a device here for multi-gpu setups

File: .gitignore [hunk 3]
```diff
@@ -2,6 +2,7 @@
 /data/
 /logs/
 /output/
+*.csv
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply .gitignore:5]: Yes, the `benchmark.py` script outputs a CSV file which we don't want to track with git.

[the remark this answers is not in the document]
Comment (calebrob6) [reply .gitignore:5]: gotcha, I was remembering our conversation about global ignores (and have since added CSVs to my global ignore)

**Review by calebrob6 — APPROVED** [review]

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply .gitignore:5]: Yes, since this file is generated by the project and not specific to a user's OS or IDE, I think this belongs in the project gitignore.
